feat(event-handler): added compress middleware for the REST API event handler

sdangol opened the pull request:
## Summary

This PR adds a built in compress middleware which can be used along with the event handler for providing an easy way to compress response bodies to reduce payload size and improve performance.

### Changes

> Please provide a summary of what's being changed

- Added a `middleware` directory in the rest directory to contain the built-in middlewares
- Exported the middleware path to be used
- Created the compress middleware which takes an encoding and a threshold and does the compression if needed

<!-- What is this PR solving? Write a clear description or reference the issue(s) it addresses. -->

> Please add the issue number below, if no issue is present the PR might get blocked and not be reviewed

**Issue number:** closes #4474 

<!-------
Before creating the pull request, please make sure you do the following:

- Read the Contributing Guidelines at https://github.com/aws-powertools/powertools-lambda-typescript/blob/main/CONTRIBUTING.md#sending-a-pull-request
- Check that there isn't already a PR that addresses the same issue. If you find a duplicate, please leave a comment under the existing PR so we can discuss how to move forward
- Check that the change meets the project's tenets https://docs.powertools.aws.dev/lambda/typescript/latest/#tenets
- Add a PR title that follows the conventional commit semantics - https://github.com/aws-powertools/powertools-lambda-typescript/blob/main/.github/semantic.yml#L2
- If relevant, add tests that prove that the change is effective and works
- Whenever relevant, make sure to comment functions/methods/types and make appropriate changes to the documentation
------->

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

dreamorosi commented on packages/event-handler/src/rest/middleware/compress.ts:
```diff
@@ -0,0 +1,58 @@
+import type { Middleware } from '@aws-lambda-powertools/event-handler/types';
```
This should be a local relative import.

The rule of thumb is:
- if it's under the same `packages/<name>/src` folder, it uses a relative import 
- if it's under a different one, it uses a full named import (like what you put above)

dreamorosi commented on packages/event-handler/src/rest/middleware/compress.ts:
```diff
@@ -0,0 +1,58 @@
+import type { Middleware } from '@aws-lambda-powertools/event-handler/types';
+import type { CompressionOptions } from 'src/types/rest.js';
```
This should also be a relative import starting with `../` - this is because the module resolution is always relative to the file that is importing, not the package root - despite what VSCode auto-suggests.

dreamorosi commented:
Code duplication is quite high, I suspect because of tests. Can we do anything to bring it down?

Also, and I suspect the answer is no, but is there any way to break down that regex into smaller chunks? Not just because Sonar complains (which I'd like to avoid if possible) but also because of complexity.

Can we, for example, make an educated guess about a subset of content types that are more common and put these in their own smaller (and faster?) regex, and then fall back to the second (or third) if the previous fail? Does this make sense?

dreamorosi commented on packages/event-handler/package.json:
```diff
@@ -77,6 +77,16 @@
         "default": "./lib/esm/rest/index.js"
       }
+    },
+    "./experimental-rest/middleware": {
```
We need to add the new path to the `typeVersions` below as well, or when you actually use it, this happens:
<img width="826" height="167" alt="Screenshot 2025-09-17 at 00 04 44" src="https://github.com/user-attachments/assets/181be4b5-6099-4229-91f2-f90a163f4fc4" />

dreamorosi commented on packages/event-handler/src/rest/middleware/compress.ts:
````diff
@@ -0,0 +1,120 @@
+ *
+ * @example
+ * ```typescript
+ * import { Router } from '@aws-lambda-powertools/event-handler';
+ * import { compress } from '@aws-lambda-powertools/event-handler/rest/middleware';
````
These paths are wrong, they should be:

```ts
import { Router } from '@aws-lambda-powertools/event-handler/experimental-rest';
import { compress } from '@aws-lambda-powertools/event-handler/experimental-rest/middleware';
```

dreamorosi commented on packages/event-handler/src/rest/middleware/compress.ts:
````diff
@@ -0,0 +1,120 @@
+ *
+ * @example
+ * ```typescript
+ * import { Router } from '@aws-lambda-powertools/event-handler';
+ * import { compress } from '@aws-lambda-powertools/event-handler/rest/middleware';
````
These package names are wrong - should be:

```ts
import { Router } from '@aws-lambda-powertools/event-handler/experimental-rest';
import { compress } from '@aws-lambda-powertools/event-handler/experimental-rest/middleware';
```

dreamorosi commented on packages/event-handler/src/rest/middleware/compress.ts:
```diff
@@ -0,0 +1,120 @@
+  const threshold = options?.threshold ?? 1024;
+
+  return async (_, reqCtx, next) => {
+    await next();
```
The [Python docs](https://docs.powertools.aws.dev/lambda/python/latest/core/event_handler/api_gateway/#compress) suggest that even with `compress` enabled, if the client doesn't send the correct `Accept-Encoding` header, Event Handler won't return a compressed response.

Can you check their implementation to see if this is this the case?

dreamorosi replied:
@svozza what do you think about this? I saw at least another JS implementation doesn't do this, but I think the client should be able to bypass the API compression setting if they don't want to/can't support it, otherwise the request would be thrown away anyway.

This also matches the behavior from API Gateway as documented [here](https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-make-request-with-compressed-payload.html) and [here](https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-gzip-compression-decompression.html).

svozza replied:
No I agree, that header is there for a reason and we should respect it.

dreamorosi commented:
I think I found the issue: at [this point](https://github.com/aws-powertools/powertools-lambda-typescript/blob/87058a29c44dff0ed0f53104b058e8c8a334ee05/packages/event-handler/src/rest/converters.ts#L92-L98) when we convert the `Response` object to an API Gateway Proxy response we hardcode both `isBase64Encoded` and `body` respectively as `false` and text.

I applied a quick & dirty patch and it works:

```ts
// Check if response contains compressed/binary content
const isCompressed = response.headers.has('content-encoding');
let body: string;
let isBase64Encoded = false;

if (isCompressed) {
  // For compressed content, get as buffer and encode to base64
  const buffer = await response.arrayBuffer();
  body = Buffer.from(buffer).toString('base64');
  isBase64Encoded = true;
} else {
  // For text content, use text()
  body = await response.text();
  isBase64Encoded = false;
}

const result: APIGatewayProxyResult = {
  statusCode: response.status,
  headers,
  body,
  isBase64Encoded,
};
```

I don't know if there's a better way to do this, or if I'm missing any edge cases, so please triple check the code above if you decide to put it in the repo.

Also I'm not sure if compression is the only reason why the `isBase64Encoded` field should be set to `true` (probably one of the main ones) - but worth checking. 

@svozza - keen to hear your take on this as well.

svozza commented:
> I think I found the issue: at [this point](https://github.com/aws-powertools/powertools-lambda-typescript/blob/87058a29c44dff0ed0f53104b058e8c8a334ee05/packages/event-handler/src/rest/converters.ts#L92-L98) when we convert the `Response` object to an API Gateway Proxy response we hardcode both `isBase64Encoded` and `body` respectively as `false` and text.
> 
> I applied a quick & dirty patch and it works:
> 
> ```ts
> // Check if response contains compressed/binary content
> const isCompressed = response.headers.has('content-encoding');
> let body: string;
> let isBase64Encoded = false;
> 
> if (isCompressed) {
>   // For compressed content, get as buffer and encode to base64
>   const buffer = await response.arrayBuffer();
>   body = Buffer.from(buffer).toString('base64');
>   isBase64Encoded = true;
> } else {
>   // For text content, use text()
>   body = await response.text();
>   isBase64Encoded = false;
> }
> 
> const result: APIGatewayProxyResult = {
>   statusCode: response.status,
>   headers,
>   body,
>   isBase64Encoded,
> };
> ```
> 
> I don't know if there's a better way to do this, or if I'm missing any edge cases, so please triple check the code above if you decide to put it in the repo.
> 
> Also I'm not sure if compression is the only reason why the `isBase64Encoded` field should be set to `true` (probably one of the main ones) - but worth checking.
> 
> @svozza - keen to hear your take on this as well.

Yes, the `base64` logic is something I [knew](https://github.com/aws-powertools/powertools-lambda-typescript/pull/4349#discussion_r2280500147) we'd have to come back to. I'm not sure if the presence of the `content-encoding` header is enough, I think we need to check for specific values in the header such as `deflate` and `gzip`. We also need to check the `content-type` headers for binary MIME types like `image/png`, `image/jpeg` etc.

dreamorosi commented:
Ok, thanks.

I think we should address that in a separate issue that possibly has its own spec with requirements and all.

For the sake of this PR, let's just make sure that when the `compress` middleware is enabled, `isBase64Encoded` is also set to `true` in the final response object.

sdangol commented:
@dreamorosi @svozza I've refactored the code a bit now.

It's now using the `accepted-encoding` header value to to find the preferred encoding and is disabled using the `identity` value.
I've removed the `content-type` checking to simplify things for now. I think if there are any requests to support additional types, we can add them later.

svozza commented on packages/event-handler/tests/unit/rest/middleware/compress.test.ts:
```diff
@@ -0,0 +1,169 @@
+import { compress } from '../../../../src/rest/middleware/index.js';
+import { createSettingHeadersMiddleware, createTestEvent } from '../helpers.js';
+
+describe('Compress Middleware', () => {
```
We should have assertions to check `isBase64Encoded` is set properly in these tests too.

sdangol commented:
> I think that's expected and the result of the content being compressed or not - I guess the only check would be to make sure it matches the expected value in the two cases.

The responses did match in both the cases

svozza commented on packages/event-handler/tests/unit/rest/converters.test.ts:
```diff
@@ -396,6 +396,18 @@ describe('Converters', () => {
+      });
+
+      const result = await webResponseToProxyResult(response);
+      expect(result.isBase64Encoded).toBe(true);
```
We need to check the body here too make sure that the logic that converts it to a base64 string works.
